plugins/language_typst.lua

```diff
@@ -4,153 +4,213 @@ local syntax = require "core.syntax"
+-- so in order to fix this issue, I had to create this extra sub-syntax.
+local typst_string = {
 
-local typst_script = {
-
-    patterns = {
-
-        { pattern = { "/%*", "%*/" },              type = "comment" },
-        { pattern = "//.*",                        type = "comment" },
-
-        { pattern = { '"', '"', '\\' },            type = "string"  },
-        { pattern = {"%[", "%]"},                  type = "normal", syntax = ".typ" },
-
-        { pattern = "[%a_][%w_%-]*%f[(]",          type = "function" },
-
-        { pattern = "0x[%dabcdef]+",               type = "number" },
-        { pattern = "0b[01]+",                     type = "number" },
-        { pattern = "0o[01234567]+",               type = "number" },
-        { pattern = "%d+[.e]%d+()[%a%%]+",         type = {"number", "keyword2"}},
-        { pattern = "%d+[.e]%d+",                  type = "number" },
-        { pattern = "%d+()[%a%%]+",                type = {"number", "keyword2"}},
-        { pattern = "%d",                          type = "number" },
+	patterns = {
+		{ pattern = "%g", type = "string" }
+	},
 
-
-        { pattern = "[%+%-=/%*%^%%<>~|&#_^\\]",    type = "operator" },
-
-    },
-
-
-    symbols = {
-        ["set"] = "keyword",
-        ["let"] = "keyword",
-        ["include"] = "keyword",
-    }
+	symbols = {}
 }
 
 
-syntax.add {
-    name = "Typst",
-    files = { "%.typ$" },
-    comment = "//",
-    block_comment = { "/*", "*/" },
-
-    patterns = {
-
-
-        { pattern = {"%$", "%$", "\\"},            type = "literal", syntax = typst_math },
-        { pattern = {"```js", "```", "\\"},        type = "literal", syntax = ".js" },
-        { pattern = {"#", "[^,%(%[{]\n"},          type = "literal", syntax = typst_script },
-        { pattern = {"`", "`", "\\"},              type = "normal"},
-
-        { pattern = { '"', '"', '\\' },            type = "string"  },
-
-
-        { pattern = "//.*",                        type = "comment" },
-        { pattern = { "/%*", "%*/" },              type = "comment" },
-
-   --     { pattern = {"link"},                    type = "typst_underline"}, -- add underline text later
-
-        { pattern = "^%s*=+ ().+%f[\n]",             type ={"operator", "literal"}}, -- Supposed to be bold
-
-    -- Finish later
-        { pattern = {"%*_", "_%*[%s,%.]"},         type = "keyword2"}, -- Bold Italic
-        { pattern = {"_%*", "%*_[%s,%.]"},         type = "keyword2"}, -- Bold Italic
-        { pattern = {"_", "_[%s,%.]"},             type = "keyword"}, -- Italic
-        { pattern = {"%*[^%/]", "%*[%s,%.]"},      type = "literal"}, -- Bold
+local typst_script = {
 
-        { pattern = "[%+%-\\]",    type = "operator" },
+	patterns = {
+
+		{ pattern = { "/%*", "%*/" },              type = "comment"                       },
+		{ pattern = "//.*",                        type = "comment"                       },
+
+		{ pattern = { '"', '"', '\\' },            type = "string", syntax = typst_string },
```

Comment:
What issue does this solve?

Reply:
There was a bug where the string highlighting did not work properly (probably because the tokenizer prioritizes subsyntaxes over classic start-end patterns). So in order to fix this issue, I had to create this extra sub-syntax.

Reply:
Could you give me an example?
Just to check if it's something we need to fix, to understand the problem better.

Reply:
Without the extra syntax:
![image](https://github.com/lite-xl/lite-xl-plugins/assets/93390411/91e33e10-c479-4c7e-8ee3-68aa5b5b6788)

With the extra syntax:
![image](https://github.com/lite-xl/lite-xl-plugins/assets/93390411/c50ed0ef-ccb5-4cbb-ab9a-c6cf16463e73)

Reply:
The issue is caused by
https://github.com/lite-xl/lite-xl-plugins/blob/553b178d929374f4bd2aba50851edbdbbd8986a9/plugins/language_typst.lua#L183

which is stopping at the first space.
Is that intended? Or should it go ahead until the newline?

Reply:
It is intended, you are supposed to do something like that:
```
#let name = "Typst"
This is #name's documentation.
It explains #name.
```

Reply:
Then it might be a good idea to provide both versions.
One for full line match, and another for "up to the fist space".
For the full line one, you should probably use `^` at the start.

Reply:
This is sort of what I did: I noticed that when it needed a full line match, it was when using keywords (as `let`, `set`, `import` etc...)
In all other cases, it stops at the first space encountered.
This works for most of the cases except for the case where you use directly a string after the `#`, which is why I had to add this extra sub-syntax.

And also, if I used `^` in order to detect a full line capture, something like this wouldn't work:

```
#let name = "Typst"
#name is great!
```

Reply:
Then I suggest using:
```lua
local typst_string = {

	patterns = {
		{ pattern = "\\.",   type = "string" },
		{ pattern = '[^"\\]+', type = "string" }
	},

	symbols = {}
}
```
This way it matches as much as possible at once, reducing the number of times it needs to run.

Reply:
Thanks! I've added it (and a similar thing for the python syntax).